torch/csrc/distributed/rpc/rref.cpp

```diff
@@ -1,119 +1,188 @@
 #include <torch/csrc/distributed/rpc/rref.h>
+
+#include <torch/csrc/distributed/rpc/python_rpc_handler.h>
 #include <torch/csrc/distributed/rpc/rref_context.h>
-#include <torch/csrc/distributed/rpc/script_rref_proto.h>
+#include <torch/csrc/distributed/rpc/rref_proto.h>
 
 namespace torch {
 namespace distributed {
 namespace rpc {
 
+namespace {
+
+constexpr int RFD_TUPLE_SIZE = 6; // number of RRefForkData fields in py::tuple
+constexpr int OWNER_IDX = 0; // index of ownerId in the tuple
+constexpr int RREFID_ON_IDX = 1; // index of RRefId.createdOn_ in the tuple
+constexpr int RREFID_ID_IDX = 2; // index of RRefId.localId_ in the tuple
+constexpr int FORKID_ON_IDX = 3; // index of ForkId.createdOn_ in the tuple
+constexpr int FORKID_ID_IDX = 4; // index of ForkId.localId_ in the tuple
+constexpr int PARENT_IDX = 5; // index of parent in the tuple
+
+} // namespace
+
 std::atomic<local_id_t> RRefContext::nextLocalId_{0};
 
 //////////////////////////  RRefForkData  /////////////////////////////////
 
 RRefForkData::RRefForkData(
     worker_id_t ownerId,
     const RRefId& rrefId,
-    const ForkId& forkId)
-    : ownerId_(ownerId), rrefId_(rrefId), forkId_(forkId) {}
+    const ForkId& forkId,
+    worker_id_t parent)
+    : ownerId_(ownerId), rrefId_(rrefId), forkId_(forkId), parent_(parent) {}
 
 at::IValue RRefForkData::toIValue() const {
-  std::vector<at::IValue> ivalues = {
-      (int64_t)ownerId_, rrefId_.toIValue(), forkId_.toIValue()};
+  return c10::ivalue::Tuple::create({static_cast<int64_t>(ownerId_),
+                                     rrefId_.toIValue(),
+                                     forkId_.toIValue(),
+                                     static_cast<int64_t>(parent_)});
+}
+
+py::tuple RRefForkData::toPyTuple() const {
+  return py::make_tuple(
+      ownerId_,
+      rrefId_.createdOn_,
+      rrefId_.localId_,
+      forkId_.createdOn_,
+      forkId_.localId_,
+      parent_);
+}
 
-  return c10::ivalue::Tuple::create(std::move(ivalues));
+RRefForkData RRefForkData::fromPyTuple(const py::tuple& t) {
+  TORCH_INTERNAL_ASSERT(
+      t.size() == RFD_TUPLE_SIZE,
+      "Pickled RRefForkData must contain 6 numbers.");
+  worker_id_t ownerId = t[OWNER_IDX].cast<worker_id_t>();
+  RRefId rrefId = RRefId(
+      t[RREFID_ON_IDX].cast<worker_id_t>(),
+      t[RREFID_ID_IDX].cast<local_id_t>());
+  RRefId forkId = RRefId(
+      t[FORKID_ON_IDX].cast<worker_id_t>(),
+      t[FORKID_ID_IDX].cast<local_id_t>());
+  worker_id_t parent = t[PARENT_IDX].cast<worker_id_t>();
+  return RRefForkData(ownerId, rrefId, forkId, parent);
```

Comment:
Per above comment about the constructor, you would be able to std::move them into here (or construct inline) and elide the copy. These are tiny structs so I doubt performance impact is measurable, but good practice nonetheless.